tests/inliner-tests.lisp

```diff
@@ -6,3 +6,106 @@
   ;; See gh #1202
   (is (== 1 (1+
              ((fn (x) (return x)) 0)))))
+
+(coalton-toplevel
+  (declare two-arg-double-float-add
+     (Double-Float -> Double-Float -> Double-Float))
+  (inline)
+  (define (two-arg-double-float-add x y)
+    (lisp Double-Float (x y) (cl:+ x y))))
+
+(coalton-toplevel
+  (define (two-arg-double-float-add-caller x y)
+    (two-arg-double-float-add x y))
+  (define two-arg-double-float-add-underapply
+    (two-arg-double-float-add 3.0d0))
+  (define (two-arg-double-float-add-underapply-caller x)
+    (two-arg-double-float-add-underapply x)))
+
+(define-test function-inline ()
+  (is (== 5.0d0 (two-arg-double-float-add-caller 2.0d0 3.0d0)))
+  (is (== 5.0d0 (two-arg-double-float-add-underapply-caller 2.0d0))))
+
+(coalton-toplevel
+  (define-class (class-for-inline-method-test :a)
+    (method-for-inline-test (:a -> :a -> :a)))
+  (define-instance (class-for-inline-method-test Double-Float)
+    (inline)
+    (define (method-for-inline-test x y)
+      (lisp Double-Float (x y) (cl:+ x y)))))
+
+(coalton-toplevel
+  (declare method-for-inline-test-caller
+    (Double-Float -> Double-Float -> Double-Float))
+  (define (method-for-inline-test-caller x y)
+    (method-for-inline-test x y)))
+
+(define-test method-inline ()
+  (is (== 5.0d0 (method-for-inline-test-caller 2.0d0 3.0d0))))
+
+(coalton-toplevel
+  (inline)
+  (define (generic-for-inline x y)
+    (+ x y))
+
+  (declare num-generic-for-inline (Num :a => :a -> :a))
+  (inline)
+  (define num-generic-for-inline
+    (generic-for-inline 2)))
+
+(coalton-toplevel
+  (monomorphize)
+  (declare monomorph-for-inline (Integer -> Integer))
+  (define (monomorph-for-inline y)
+    (num-generic-for-inline y)))
+
+(define-test monomorphize-inline ()
+  (is (== 5 (monomorph-for-inline 3))))
+
+
+(in-package #:coalton-tests)
+
+(deftest function-inline-error ()
+  (signals coalton-impl/parser:parse-error
+    (check-coalton-types
+     "(inline myfun)
+      (define (myfun x) x)")))
+
+(deftest method-inline-error ()
+  (signals coalton-impl/parser:parse-error
+    (check-coalton-types
+     "(define-class (C :a)
+        (m (:a -> :a -> :a)))
+      (define-instance (m Double-Float)
+        (inline m)
+        (define (m x y)
+          (lisp Double-Float (x y) (cl:+ x y))))")))
```

Comment:
You could add another test for duplicate inline attributes to match the tests for `monomorphize` ( https://github.com/search?q=repo%3Acoalton-lang%2Fcoalton+monomorphize+path%3A%2F%5Etests%5C%2Ftest-files%5C%2F%2F&type=code ) but other than that I think this looks pretty good and complete

Reply:
Should the tests be only for the duplicate inline attributes, or for others too (define, parse-toplevel, parse-attribute, define-type)?

Reply:
I would think that adding more tests is never a bad thing, especially if you think they would be useful

Reply:
Added in https://github.com/coalton-lang/coalton/pull/1290/commits/7d98dcaabce6a918d28ed2af8394629af059a20e and https://github.com/coalton-lang/coalton/pull/1290/commits/9edb4b3b1167fc461ddfa23126b8e2e5544bb0dc